File: simple64-updater.py
#!/usr/bin/python3

import threading
import subprocess  # nosec
import requests
import tempfile
import sys
import os
import zipfile
import shutil
import tkinter as tk


def clean_dir(install_path: str) -> None:
    # loop through the directory twice, one to clean up files, then again to remove empty dirs
    for root, _, files in os.walk(install_path):
        for name in files:
            path = os.path.join(root, name)
            if path.endswith(".dll") or path.endswith(".exe"):
                try:
                    os.remove(path)
                except OSError:
                    pass

    for root, dirs, _ in os.walk(install_path, topdown=False):  # remove empty dirs
        for name in dirs:
            try:
                os.rmdir(os.path.join(root, name))
            except OSError:
                pass


def update_simple64(root2: tk.Tk, var2: tk.StringVar) -> None:
    var2.set("Determining latest release")
    resp = requests.get(
        "https://api.github.com/repos/simple64/simple64/releases/latest", timeout=30
    )
    if resp.status_code != 200:
        root2.quit()
        return
    simple64_url = ""
    for item in resp.json()["assets"]:
        if sys.platform.startswith("win") and "simple64-win64" in item["name"]:
            simple64_url = item["browser_download_url"]

    if not simple64_url:  # quit early if URL couldn't be determined
        root2.quit()
        return

    var2.set("Downloading latest release")
    resp = requests.get(simple64_url, allow_redirects=True, timeout=30)
    if resp.status_code != 200:
        root2.quit()
        return

    clean_dir(sys.argv[1])  # clean up current directory

    with tempfile.TemporaryDirectory() as tempdir:
        filename = os.path.join(tempdir, "simple64.zip")
        with open(filename, "wb") as localfile:
            localfile.write(resp.content)  # write zip into temp dir

        var2.set("Extracting release")
        with zipfile.ZipFile(filename, "r") as zf:
            for info in zf.infolist():
                zf.extract(info.filename, path=tempdir)
                out_path = os.path.join(tempdir, info.filename)
                perm = info.external_attr >> 16
                try:
                    os.chmod(out_path, perm)
                except OSError:
                    pass

        var2.set("Moving files into place")
        extract_path = os.path.join(tempdir, "simple64")
        shutil.copytree(extract_path, sys.argv[1], dirs_exist_ok=True)

    var2.set("Cleaning up")
    root2.quit()


def start_thread(x2: threading.Thread) -> None:
    x2.start()


def main() -> None:
    if len(sys.argv) < 2:
        print("no argument!")
        sys.exit(1)

    my_env = os.environ.copy()

    root = tk.Tk()
    root.geometry("400x200")
    root.title("simple64-updater")
    var = tk.StringVar()
    var.set("Initializing")
    w = tk.Label(root, textvariable=var)
    w.pack(fill="none", expand=True)

    x = threading.Thread(target=update_simple64, args=(root, var))
    root.after(3000, start_thread, x)
    root.mainloop()
    x.join()

    subprocess.Popen(  # nosec
        [os.path.join(sys.argv[1], "simple64-gui")],
        env=my_env,
        start_new_session=True,
        close_fds=True,
    )


if __name__ == "__main__":
    main()

File: simple64-updater.go
package main

import (
	"archive/zip"
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/layout"
	"fyne.io/fyne/v2/widget"
	retryablehttp "github.com/hashicorp/go-retryablehttp"
)

func printError(label *widget.Label, app fyne.App, s string) {
	label.SetText(s)
	time.Sleep(3 * time.Second) //nolint:gomnd,mnd
	app.Quit()
}

func cleanDir(directory string) error {
	err := filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if info.IsDir() {
			return nil // Skip directories
		}

		// Check if the file has the desired extension
		if strings.HasSuffix(info.Name(), ".exe") || strings.HasSuffix(info.Name(), ".dll") {
			if err = os.Remove(path); err != nil {
				return fmt.Errorf("could not remove file: %s", err.Error())
			}
		}

		return nil
	})
	if err != nil {
		return fmt.Errorf("could not clean directory: %s", err.Error())
	}
	return nil
}

func determineLatestRelease(label *widget.Label) (string, error) {
	label.SetText("Determining latest release")
	httpClient := retryablehttp.NewClient()
	httpClient.Logger = nil
	resp, err := httpClient.Get("https://api.github.com/repos/simple64/simple64/releases/latest")
	if err != nil {
		return "", fmt.Errorf("error determining latest release: %s", err.Error())
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("error determining latest release, http status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("could not read HTTP response: %s", err.Error())
	}

	var data map[string]interface{}
	if err = json.Unmarshal(body, &data); err != nil {
		return "", fmt.Errorf("error parsing JSON: %s", err.Error())
	}
	simple64Url := ""
	assets, ok := data["assets"].([]interface{})
	if !ok {
		return "", fmt.Errorf("error reading assets")
	}
	for _, element := range assets {
		subArray, ok := element.(map[string]interface{})
		if !ok {
			return "", fmt.Errorf("error reading assets")
		}
		name, ok := subArray["name"].(string)
		if !ok {
			return "", fmt.Errorf("error asset file name")
		}
		if strings.Contains(name, "simple64-win64") {
			simple64Url, ok = subArray["browser_download_url"].(string)
			if !ok {
				return "", fmt.Errorf("error reading download url")
			}
		}
	}

	if simple64Url == "" {
		return simple64Url, fmt.Errorf("could not determine download URL")
	}
	return simple64Url, nil
}

func downloadRelease(simple64Url string, label *widget.Label) ([]byte, int64, error) {
	label.SetText("Downloading latest release")
	httpClient := retryablehttp.NewClient()
	httpClient.Logger = nil
	zipResp, err := httpClient.Get(simple64Url)
	if err != nil {
		return nil, 0, fmt.Errorf("error downloading latest release: %s", err.Error())
	}
	defer zipResp.Body.Close()

	if zipResp.StatusCode != http.StatusOK {
		return nil, 0, fmt.Errorf("error downloading latest release, http status: %d", zipResp.StatusCode)
	}

	zipBody, err := io.ReadAll(zipResp.Body)
	if err != nil {
		return nil, 0, fmt.Errorf("could not read HTTP response: %s", err.Error())
	}
	return zipBody, zipResp.ContentLength, nil
}

func prepDirectory(label *widget.Label) error {
	label.SetText("Cleaning existing directory")

	// Create the output directory if it doesn't exist
	if err := os.MkdirAll(os.Args[1], os.ModePerm); err != nil {
		return fmt.Errorf("could not create directory: %s", err.Error())
	}

	if err := cleanDir(os.Args[1]); err != nil {
		return fmt.Errorf("could not clean existing directory: %s", err.Error())
	}
	return nil
}

func extractZip(label *widget.Label, zipBody []byte, zipLength int64) error {
	label.SetText("Extracting ZIP archive")

	// Open the downloaded zip file
	zipReader, err := zip.NewReader(bytes.NewReader(zipBody), zipLength)
	if err != nil {
		return fmt.Errorf("could not open ZIP: %s", err.Error())
	}

	// Extract each file from the zip archive
	for _, file := range zipReader.File {
		// Open the file from the archive
		zipFile, err := file.Open()
		if err != nil {
			return fmt.Errorf("could not open ZIP file: %s", err.Error())
		}
		defer zipFile.Close()

		if !file.FileInfo().IsDir() {
			// Construct the output file path
			relPath, err := filepath.Rel("simple64", file.Name)
			if err != nil {
				return fmt.Errorf("could not determine file path: %s", err.Error())
			}
			outputPath := filepath.Join(os.Args[1], relPath)

			// Create the parent directory of the file
			if err = os.MkdirAll(filepath.Dir(outputPath), os.ModePerm); err != nil {
				return fmt.Errorf("could not create directory: %s", err.Error())
			}

			// Create the output file
			outputFile, err := os.OpenFile(outputPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, file.Mode())
			if err != nil {
				return fmt.Errorf("could not create file: %s", err.Error())
			}
			defer outputFile.Close()

			// Copy the contents from the zip file to the output file
			for {
				_, err := io.CopyN(outputFile, zipFile, 1024) //nolint:gomnd,mnd
				if err != nil {
					if errors.Is(err, io.EOF) {
						break
					}
					return fmt.Errorf("could not copy file: %s", err.Error())
				}
			}
		}
	}
	return nil
}

func updateSimple64(label *widget.Label, app fyne.App, c chan bool) {
	time.Sleep(3 * time.Second) //nolint:gomnd,mnd // Wait for simple64-gui to close

	simple64Url, err := determineLatestRelease(label)
	if err != nil {
		printError(label, app, err.Error())
		c <- false
		return
	}

	zipBody, zipLength, err := downloadRelease(simple64Url, label)
	if err != nil {
		printError(label, app, err.Error())
		c <- false
		return
	}

	if err = prepDirectory(label); err != nil {
		printError(label, app, err.Error())
		c <- false
		return
	}

	if err = extractZip(label, zipBody, zipLength); err != nil {
		printError(label, app, err.Error())
		c <- false
		return
	}

	label.SetText("Done extracting ZIP archive")
	time.Sleep(1 * time.Second)
	app.Quit()
	c <- true
}

func main() {
	if len(os.Args) < 2 { //nolint:gomnd,mnd
		log.Fatal("must specify target directory")
	}
	a := app.New()
	w := a.NewWindow("simple64-updater")
	w.Resize(fyne.NewSize(400, 200)) //nolint:gomnd,mnd
	label := widget.NewLabel("Initializing")
	content := container.New(layout.NewCenterLayout(), label)

	w.SetContent(content)

	c := make(chan bool)
	go updateSimple64(label, a, c)
	w.ShowAndRun()

	if <-c {
		cmd := exec.Command(filepath.Join(os.Args[1], "simple64-gui")) //nolint:gosec
		if err := cmd.Start(); err != nil {
			log.Fatal(err)
		}
		if err := cmd.Process.Release(); err != nil {
			log.Fatal(err)
		}
	}
}
